pkg/nsx/services/securitypolicy/store.go

```diff
@@ -1,153 +1,117 @@
-}
-
-func queryGroup(service *SecurityPolicyService, wg *sync.WaitGroup, fatalErrors chan error) {
-	defer wg.Done()
-	queryParam := fmt.Sprintf("%s:%s", common.ResourceType, common.ResourceTypeGroup) + " AND " + queryTagCondition(service)
-	var cursor *string = nil
-	pageSize := common.PageSize
-	for {
-		response, err := service.NSXClient.QueryClient.List(queryParam, cursor, nil, &pageSize, nil, nil)
-		err = common.TransError(err)
-		if _, ok := err.(nsxutil.PageMaxError); ok == true {
-			common.DecrementPageSize(&pageSize)
-			continue
-		}
-		if err != nil {
-			fatalErrors <- err
-		}
-		for _, g := range response.Results {
-			a, err := Converter.ConvertToGolang(g, model.GroupBindingType())
-			if err != nil {
-				for _, e := range err {
-					fatalErrors <- e
-				}
-			}
-			c, _ := a.(model.Group)
-			err2 := service.GroupStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
-		}
-		cursor = response.Cursor
-		if cursor == nil {
-			break
-		}
-		c, _ := strconv.Atoi(*cursor)
-		if int64(c) >= *response.ResultCount {
-			break
-		}
-	}
+var (
+	key = common.TagScopeSecurityPolicyCRUID
+)
+
+type resourceAssertion func(i interface{}) interface{}
+
+func keyFuncSecurityPolicy(obj interface{}) (string, error) {
+	o := obj.(model.SecurityPolicy)
+	return *o.Id, nil
 }
 
-func querySecurityPolicy(service *SecurityPolicyService, wg *sync.WaitGroup, fatalErrors chan error) {
-	defer wg.Done()
-	queryParam := fmt.Sprintf("%s:%s", common.ResourceType, common.ResourceTypeSecurityPolicy) + " AND " + queryTagCondition(service)
-	var cursor *string = nil
-	pageSize := common.PageSize
-	for {
-		response, err := service.NSXClient.QueryClient.List(queryParam, cursor, nil, &pageSize, nil, nil)
-		err = common.TransError(err)
-		if _, ok := err.(nsxutil.PageMaxError); ok == true {
-			common.DecrementPageSize(&pageSize)
-			continue
-		}
-		if err != nil {
-			fatalErrors <- err
-		}
-		for _, g := range response.Results {
-			a, err := Converter.ConvertToGolang(g, model.SecurityPolicyBindingType())
-			if err != nil {
-				for _, e := range err {
-					fatalErrors <- e
-				}
-			}
-			c, _ := a.(model.SecurityPolicy)
-			err2 := service.SecurityPolicyStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
-		}
-		cursor = response.Cursor
-		if cursor == nil {
-			break
-		}
-		c, _ := strconv.Atoi(*cursor)
-		if int64(c) >= *response.ResultCount {
-			break
+func keyFuncGroup(obj interface{}) (string, error) {
+	o := obj.(model.Group)
+	return *o.Id, nil
+}
+
+func keyFuncRule(obj interface{}) (string, error) {
+	o := obj.(model.Rule)
+	return *o.Id, nil
+}
+
+func indexFuncSecurityPolicy(obj interface{}) ([]string, error) {
+	o := obj.(model.SecurityPolicy)
+	return filterTag(o.Tags), nil
+}
+
+func indexFuncGroup(obj interface{}) ([]string, error) {
+	o := obj.(model.Group)
+	return filterTag(o.Tags), nil
+}
+
+func indexFuncRule(obj interface{}) ([]string, error) {
+	o := obj.(model.Rule)
+	return filterTag(o.Tags), nil
+}
+
```

Comment:
For these keyFunc and indexFunc, we use one  common to retrieve specific function in the origin design. 
// IndexFunc is wrapper of cache.IndexFunc, used by cache.Indexers
const (
func IndexFunc(scope string) cache.IndexFunc {
	PageSize int64 = 1000
	return func(obj interface{}) ([]string, error) {
)
		return CRUIDScopeIndexFunc(scope, obj)
	}
}
// KeyFunc returns the key of the object.
func KeyFunc(obj interface{}) (string, error) {
	switch v := obj.(type) {
	case model.Group:
		return *v.Id, nil
	case model.SecurityPolicy:
		return *v.Id, nil
	case model.Rule:
		return *v.Id, nil
	}
	return "", nil
}

I feel the original way is better than current refactored one. Current approach bring more funcs and more codes.
Also, we already have common.ResourceTypeSecurityPolicy, common.ResourceTypeRule and common.ResourceTypeGroup as resource type identifier.  we could follow this way to return key and rule functions. We see the similar usage that uses resource type as index for store's InitializeStore func like:
service.Store[ResourceTypeSecurityPolicy] = cache.NewIndexer(keyFuncSecurityPolicy, cache.Indexers{key: indexFuncSecurityPolicy})

Reply:
ok